File: tree/Construct_BST_from_given_preorder/Method2/construct.cpp
#include <bits/stdc++.h> 
using namespace std ; 

class Node{
    public:
        int data ;
        Node* right ; 
        Node* left ; 
}; 

Node* newNode(int data){
    Node* temp = new Node() ; 

    temp->data = data ; 
    temp->right = NULL ; 
    temp->left = NULL ; 

    return temp ; 
}

void printInorder(Node* node){
    if(node == NULL){
        return ; 
    }
    printInorder(node->left) ; 
    cout<<node->data<<" " ; 
    printInorder(node->right) ; 
}

Node* constructTreeUtil(int pre[], int* preIndex, int key, int min, int max, int size){
    if(*preIndex > size){
        return NULL ;
    }
    Node* root = NULL ; 
    if(key < max && key > min){
        root = newNode(key) ; 
        *preIndex = *preIndex + 1 ;

        if(*preIndex < size){
            root->left = constructTreeUtil(pre,preIndex, pre[*preIndex],min,key,size) ; 
        }

        if(*preIndex < size){
            root->right = constructTreeUtil(pre,preIndex,pre[*preIndex],key, max, size) ; 
        }
    }

    return root ; 
}

Node* constructTree(int pre[],int size){
    int preIndex = 0 ; 
    return constructTreeUtil(pre,&preIndex,pre[0],INT_MIN,INT_MAX,size) ; 
}

int main() {
    int pre[] = {10,5,1,7,40,50} ; 
    int size = sizeof(pre)/sizeof(pre[0]); 

    Node* root = constructTree(pre,size) ; 
    cout<<"In_order traversal of the constructed tree : "<<endl;
    printInorder(root) ; 

    cout<<"\n" ; 
    return 0 ; 
}

File: tree/Construct_BST_from_given_preorder/Method2/construct.py
INT_MIN = -float("inf") 
INT_MAX = float("inf")

class Node:
    def __init__(self, data):
        self.data = data 
        self.right = None 
        self.left = None 

def getPreIndex():
    return constructTreeUtil.preIndex 

def incrementPreIndex():
    constructTreeUtil.preIndex = constructTreeUtil.preIndex + 1 

def printInorder(node):
    if node is None :
        return 
    printInorder(node.left) 
    print(node.data, end = " ")
    printInorder(node.right)

def constructTreeUtil(pre, key, min, max, size):
    if (getPreIndex() > size):
        return None 

    root = None 

    if(key < max and key > min):
        root = Node(key) ; 
        incrementPreIndex()

        if(getPreIndex() < size):
            root.left = constructTreeUtil(pre, pre[getPreIndex()], min, key, size) 

        if getPreIndex() < size :
            root.right = constructTreeUtil(pre, pre[getPreIndex()], key, max, size) 


    return root 

def constructTree(pre):
    constructTreeUtil.preIndex = 0 ; 
    size = len(pre) ; 
    return constructTreeUtil(pre, pre[0], INT_MIN, INT_MAX, size)

pre = [10,5,1,7,40,50]

node = constructTree(pre)

print("In_order traversal of the constructed tree : ") 
printInorder(node) 